sa/sa.go

```diff
@@ -950,47 +950,51 @@ func (ssa *SQLStorageAuthority) leaseOldestCRLShard(ctx context.Context, req *sa
 			&shards,
 			`SELECT id, issuerID, idx, thisUpdate, nextUpdate, leasedUntil
 				FROM crlShards
-				WHERE issuerID = ?`,
-			req.IssuerNameID,
+				WHERE issuerID = ?
+				AND idx BETWEEN ? AND ?`,
+			req.IssuerNameID, req.MinShardIdx, req.MaxShardIdx,
 		)
 		if err != nil {
 			return -1, fmt.Errorf("selecting candidate shards: %w", err)
 		}
 
-		// Convert the slice to a map to detect shards that we expect to exist, but
-		// don't. At the same time, determine the oldest known shard.
-		shardMap := make(map[int]*crlShardModel, len(shards))
-		var oldest *crlShardModel
-		for _, shard := range shards {
-			if shard.Idx < int(req.MinShardIdx) || shard.Idx > int(req.MaxShardIdx) {
-				continue
+		// Determine which shard index we want to lease.
+		var shardIdx int
+		var needToInsert bool
+		if len(shards) < (int(req.MaxShardIdx + 1 - req.MinShardIdx)) {
+			// Some expected shards are missing (i.e. never-before-produced), so we
+			// pick one at random.
+			missing := make(map[int]struct{}, req.MaxShardIdx+1-req.MinShardIdx)
+			for i := req.MinShardIdx; i <= req.MaxShardIdx; i++ {
+				missing[int(i)] = struct{}{}
 			}
-			shardMap[shard.Idx] = shard
-			if shard.LeasedUntil.After(ssa.clk.Now()) {
-				continue
+			for _, shard := range shards {
+				delete(missing, shard.Idx)
 			}
-			if oldest == nil ||
-				(shard.ThisUpdate == nil && oldest.ThisUpdate != nil) ||
-				shard.ThisUpdate.Before(*oldest.ThisUpdate) {
-				oldest = shard
-			}
-		}
-
-		if oldest == nil {
-			return -1, fmt.Errorf("issuer %d has no unleased shards in range %d - %d", req.IssuerNameID, req.MinShardIdx, req.MaxShardIdx)
-		}
-
-		// Determine which shard index we want to lease: by default the oldest, but
-		// an arbitrary missing one if any are missing.
-		shardIdx := oldest.Idx
-		needToInsert := false
-		for i := req.MaxShardIdx; i >= req.MinShardIdx; i-- {
-			_, ok := shardMap[int(i)]
-			if !ok {
-				shardIdx = int(i)
-				needToInsert = true
+			for idx := range missing {
+				// Go map iteration is guaranteed to be in randomized key order.
+				shardIdx = idx
 				break
 			}
+			needToInsert = true
+		} else {
+			// We got all the shards we expect, so we pick the oldest unleased shard.
+			var oldest *crlShardModel
+			for _, shard := range shards {
+				if shard.LeasedUntil.After(ssa.clk.Now()) {
+					continue
+				}
+				if oldest == nil ||
+					(shard.ThisUpdate == nil && oldest.ThisUpdate != nil) ||
```

Comment:
Due to the line breaks and indentation, allows for a little bit easier visual diffing of the conditionals.
```suggestion
					(oldest.ThisUpdate != nil && shard.ThisUpdate.Before(*oldest.ThisUpdate)) {
```